### PR TITLE
Minor Bugs Fix

### PR DESCRIPTION
**Pull request by KapilTR**
To review:
- Added restore-wallet functionality
- refactored error log handling
- fix start-nodes bug for linux machine

### CODE REVIEW

File: modules/accounts/accounts.js [hunk 1]
```diff
@@ -111,16 +120,30 @@ class Accounts{
...
             return;
         }
 
         try {
             const balance = await RpcRequest.fetchBalance(tezosNode, pkh);
             Logger.info(Helper.formatTez(balance));  
         } catch(error) {
-            Logger.error(`${error}`);
+            if(error.toString().includes(`connect ECONNREFUSED`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Make sure local nodes are in running state....`);
```
Comment (jvtezsure) [inline modules/accounts/accounts.js]: ECONNREFUSED is a generic error and will be thrown for the remote nodes as well when you're not able to reach the remote node.

File: modules/accounts/accounts.js [hunk 2]
```diff
@@ -111,16 +120,30 @@ class Accounts{
...
+            } else if(error.toString().includes(`Unexpected end of JSON input`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Make sure account '${account}' is activated on the current provider....`);
+            } else if(error.toString().includes(`Only absolute URLs are supported`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else if(error.toString().includes(`getaddrinfo ENOTFOUND`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else if(error.toString().includes(`Invalid`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else {
+                Logger.error(`Error occured while fetching balance:\n${error}`);
```
Comment (jvtezsure) [inline modules/accounts/accounts.js]: Will this error be logged in the file or not?

File: modules/accounts/accounts.js [hunk 3]
```diff
@@ -214,7 +257,13 @@ class Accounts{
...
             Logger.info(`Testnet faucet account successfully activated: ${keys.label} - ${keys.pkh} \nWith tx hash: ${JSON.stringify(revealResult.operationGroupID)}`);
         } catch(error) {
-            Logger.error(`Error occured while activating account : ${error}`);
+            if(error.toString().includes(`Only absolute URLs are supported`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else if(error.toString().includes(`getaddrinfo ENOTFOUND`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else {
+                Logger.error(`Error occured while activating account:\n${error}`);
+            }
```
Comment (jvtezsure) [inline modules/accounts/accounts.js]: Some part of the code block is same as L140. Try to create a common method to avoid duplication.

File: modules/contract/contracts.js [hunk 4]
```diff
@@ -146,8 +146,20 @@ class Contracts {
             Logger.error(`Contract deployment has failed : ${JSON.stringify(result)}`);
             return;
         } catch(error) {
-            let parseError = `${error}`.indexOf('Instead, ');
-            Logger.error(`${error}`.substring(0, parseError != -1  ? parseError : `${error}`.length));
+            if(error.toString().includes('Unexpected word token')) {
+                let parseError = `${error}`.indexOf('Instead, ');
+                Logger.error(`${error}`.substring(0, parseError != -1  ? parseError : `${error}`.length));
+            } else if(error.toString().includes(`empty_implicit_contract`)) {
```
Comment (jvtezsure) [inline modules/contract/contracts.js]: emplty_implicit_contract means the account is not yet revelaed on the blockchain. We can reveal the account by sending some tezos to it. Try to modify the message accordingly.

File: modules/contract/contracts.js [hunk 5]
```diff
@@ -203,8 +215,22 @@ class Contracts {
...
+                Helper.errorLogHandler(`Error occured while calling the contract: ${error}`, `please wait.... contract '${contractAddress}' might take some time to get deployed on the tezos network`);;
+            } else if(error.toString().includes(`connect ECONNREFUSED`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Make sure local nodes are in running state....`);
+            } else if(error.toString().includes(`Only absolute URLs are supported`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else if(error.toString().includes(`getaddrinfo ENOTFOUND`)) {
+                Helper.errorLogHandler(`Error occured while fetching balance: ${error}`, `Current provider URL is not supported by network provider....`);
+            } else {
+                Logger.error(`Error occured while calling the contract:\n${error}`);
+            }
```
Comment (jvtezsure) [inline modules/contract/contracts.js]: Duplication of the code blocks should be avoided at any cost.

File: modules/helper.js [hunk 6]
```diff
@@ -27,6 +32,31 @@ class Helper {
...
+        }
+        while(i<config.accounts.length) {
+            if(config.accounts[i].label.includes('localnode') && !config.accounts[i].label.match(/bootstrap[1-5]/)) {
+                config.accounts.splice(i, 1);
+                config.identities.splice(i, 1);
+            } else {
+                i++;
+            }
+        }
+        jsonfile.writeFile(confFile, config);
```
Comment (jvtezsure) [inline modules/helper.js]: Have you properly tested this code block.
you initialise i with 0 and use it for the first while(){...} block. 
Before 2nd while(){...} block, you don't reset i and the value will be same as updated by first block. 
This may cause unwanted results.
use forEach() or forloop.
also avoid varialbe namong such as i,j,k. Use meaningful variable names.

File: modules/setup/setup.js [hunk 7]
```diff
@@ -36,67 +37,58 @@ class Setup {
...
-                                Logger.info('Nodes are already running....');
-                                Logger.warn(`If still facing issue run, 'tezster get-logs' or try restarting the nodes after running 'tezster stop-nodes'....`);
-                            } else {
-                                self.startNodesProgressBar();
-                            }
-                        });
+                        this.startExistingContainer();
                     } else if (error) {
-                        Logger.verbose(`Error while starting nodes error: ${error}`);
+                        Helper.errorLogHandler(`Error occured while starting the nodes ${error}`, `Error while starting the nodes, restart the nodes after running 'tezster stop-nodes'`);
```
Comment (jvtezsure) [inline modules/setup/setup.js]: Improve the 'communicatio' to user.
It should be 'error occurred while' and not 'error while' 
fix the spelling of 'occurred'
What does it mean by 'Error while starting the nodes, restart the nodes after running 'tezster stop-nodes' .

File: modules/setup/setup.js [hunk 8]
```diff
@@ -36,67 +37,58 @@ class Setup {
...
+            }
         });
     }
 
     stopNodes() {
         Logger.verbose('Command : tezster stop-nodes');
-        child_process.exec(`docker ps -a -q --format {{.Image}}`,(error, _stdout, _stderr) => {
-            if (_stdout.includes(`${IMAGE_TAG}\n`)) {
+        child_process.exec(`docker ps -a -q --format {{.Names}}`,(error, _stdout, _stderr) => {
+            if (_stdout.includes(`${CONTAINER_NAME}\n`)) {
```
Comment (jvtezsure) [inline modules/setup/setup.js]: What does it mean by 'Names'?

File: modules/setup/setup.js [hunk 9]
```diff
@@ -108,13 +100,12 @@ class Setup {
                     stream.on('end', () => {
                         writeStream.end();
                         writeStream.close();
-                        Logger.info(`log files stored as archieve format at: \n'/tmp/tezster-logs/tezster-logs.tar.gz'`);
+                        Logger.info(`Log files has been stored at following folder location as archieve format: \n'/tmp/tezster-logs/tezster-node-logs.tar.gz'`);
+                        Logger.warn(`To unzip file, run 'tar -xf tezster-node-logs.tar.gz' inside the file location folder.`);
```
Comment (jvtezsure) [inline modules/setup/setup.js]: you can remove 'inside the file location....'

File: modules/setup/setup.js [hunk 10]
```diff
@@ -142,9 +132,11 @@ class Setup {
...
                     child_process.exec(`sudo chmod -R 777 ${__dirname}/../../config.json`);
                 }
-            } else {
-                Logger.verbose(`Error in check permission while starting nodes _stderr: ${_stderr}`);
-                Logger.verbose(`Error in check permission while pulling docker image: ${error}`);
+            } else if(error){
+                Helper.errorLogHandler(`Error while changing config.json file permission: ${error}`, 'Error occured while setting up node due to permission issue');
+            } 
+            else {
+                Helper.errorLogHandler(`Error while changing config.json file permission: ${_stderr}`, 'Error occured while setting up node due to permission issue');
```
Comment (jvtezsure) [inline modules/setup/setup.js]: This both code blocks are exactly same. 
It can be simplified as 
error = error || _strderr . Also dont use underscore in variable names as our codebase doesn't follow that style.

File: modules/setup/setup.js [hunk 11]
```diff
@@ -193,6 +184,31 @@ class Setup {
...
+                        Logger.info('Nodes are in running state....');
+                    } else {
+                        Logger.error (`Error occured in previously started nodes, restart nodes after running 'tezster stop-nodes'.`);
+                    }
+                } catch (error) {
+                    Helper.errorLogHandler(`Error occured while starting the nodes${error}`, `Error occured in previously started nodes, restart nodes after running 'tezster stop-nodes'.`);
+                }
+
+                Logger.warn(`If facing any issue, fetch logs on your system using 'tezster get-logs' or try restarting the nodes after running 'tezster stop-nodes'.`);
+
```
Comment (jvtezsure) [inline modules/setup/setup.js]: If there's error, why do you need to perform 'checking node-status' flow?
It doesn;t make sense.

File: package.json [hunk 12]
```diff
@@ -55,5 +39,21 @@
     "winston": "^3.2.1",
     "xmlhttprequest": "^1.8.0",
     "zxcvbn": "^4.4.2"
-  }
+  },
```
Comment (jvtezsure) [inline package.json:42]: Always merge from dev branch first and then raise a PR to avoid unnecessary changes to dev branch,

**Review by jvtezsure — CHANGES_REQUESTED** [review]

File: modules/accounts/accounts.js (hunk 2 above)
Comment (KapilTR) [reply modules/accounts/accounts.js]: yes this error will be logged in file. "Helper.errorLogHandler" is being used to store and actual error in file and formatting the error for console so that user will get understandable format of the error. "Logger.error" will also be redirected to the file.

File: modules/setup/setup.js (hunk 8 above)
Comment (KapilTR) [reply modules/setup/setup.js]: 'Names' is the container name which is - 'tezster' in our scenario. And 'Names' is standard column name while listing all containers.

File: modules/setup/setup.js (hunk 11 above)
Comment (KapilTR) [reply modules/setup/setup.js]: In case, user ran start-nodes, and nodes didn't started successfully but container will be up. I am checking for if container is up and nodes didn't get started, we will show error accordingly. This is the possibility scenario, hence I added that.

File: package.json (hunk 12 above)
Comment (KapilTR) [reply package.json:42]: I merged from dev branch first, but that extra line space came by mistake by me. I am always aware to avoid unnecessary changes, but my bad this one came into the picture. Will not happen again.

File: modules/accounts/accounts.js [hunk 13]
```diff
@@ -39,8 +40,8 @@ class Accounts{
     async listAccounts() {  
         Logger.verbose(`Command : tezster list-accounts`);
         if(Object.keys(config.accounts).length > 0) {
-            for(var i in config.accounts) {
-                Logger.info(config.accounts[i].label + ' - ' + config.accounts[i].pkh + '(' + config.accounts[i].identity + ')');
+            for(var objIndex in config.accounts) {
```
Comment (jvtezsure) [inline modules/accounts/accounts.js]: use 'let' and 
variable can be named as 'accountIndex'.
if you use forEach, you don't have to deal with indexes.

File: modules/accounts/accounts.js [hunk 14]
```diff
@@ -227,26 +259,22 @@ class Accounts{
...
 
         try {
-            for(var i=0;i<config.accounts.length;i++) {
-                if(config.accounts[i].identity === account  || config.accounts[i].label === account || config.accounts[i].pkh === account) {
-                    config.accounts.splice(i, 1);
-                }
-            }
-            for(var i=0;i<config.identities.length;i++) {
-                if(config.identities[i].pkh === account  || config.identities[i].label === account) {
+            for(var objIndex=0; objIndex<config.accounts.length; objIndex++) {
```
Comment (jvtezsure) [inline modules/accounts/accounts.js]: you can just name to index or accountIndex.

File: modules/exceptionHandler.js [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        }
+        else if(error.toString().includes(`Only absolute URLs are supported`)) {
+            Helper.errorLogHandler(`Error occurred while ${commandType}: ${error}`, `Current provider URL is not supported by network provider....`);
+        } 
+        else if(error.toString().includes(`getaddrinfo ENOTFOUND`)) {
+            Helper.errorLogHandler(`Error occurred while ${commandType}: ${error}`, `Current provider URL is not supported by network provider....`);
+        } 
+        else if(error.toString().includes(`Invalid`)) {
+            Helper.errorLogHandler(`Error occurred while ${commandType}: ${error}`, `Current provider URL is not supported by network provider....`);
+        }
```
Comment (jvtezsure) [inline modules/exceptionHandler.js]: you can create constant for each error message, 
this way it'll be reusable and also easy to modify.

File: modules/helper.js [hunk 16]
```diff
@@ -27,6 +32,30 @@ class Helper {
...
+        let contractObjectIndex , accountObjectIndex;
+        for (contractObjectIndex=0; contractObjectIndex<config.contracts.length; contractObjectIndex++) {
+            if(config.contracts[contractObjectIndex].identity.includes('localnode')) {
+                config.contracts.splice(contractObjectIndex, 1);
+                contractObjectIndex--;
+            }
+        }
+
+        for (accountObjectIndex=0; accountObjectIndex<config.accounts.length; accountObjectIndex++) {
+            if(config.accounts[accountObjectIndex].label.includes('localnode') && !config.accounts[accountObjectIndex].label.match(/bootstrap[1-5]/)) {
```
Comment (jvtezsure) [inline modules/helper.js]: use 2 lines for if condition if it's too long.

File: modules/setup/setup.js [hunk 17]
```diff
@@ -16,87 +17,74 @@ class Setup {
...
+                Helper.errorLogHandler(`Error occurred while starting the nodes: ${error}`, `No inbuilt nodes found on system. Run 'tezster setup' command for build the nodes.`);
+            }
         });
     }
 
     stopNodes() {
         Logger.verbose('Command : tezster stop-nodes');
-        child_process.exec(`docker ps -a -q --format {{.Image}}`,(error, _stdout, _stderr) => {
-            if (_stdout.includes(`${IMAGE_TAG}\n`)) {
+        child_process.exec(`docker ps -a -q --format {{.Names}}`,(error, stdout, stderr) => {
```
Comment (jvtezsure) [inline modules/setup/setup.js:56]: Why a variable is named as 'NAMES' ?

**Review by jvtezsure — CHANGES_REQUESTED** [review]

File: modules/setup/setup.js (hunk 17 above)
Comment (KapilTR) [reply modules/setup/setup.js:56]: We didn't named this, it is docker nomenclature. We can see container name column named as 'NAMES'.
